File: src/main/java/com/myapp/controller/RelationshipController.java
package com.myapp.controller;

import com.myapp.service.RelationshipService;
import com.myapp.service.exceptions.RelationshipNotFoundException;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.web.bind.annotation.*;

@RestController
@RequestMapping("/api/relationships")
public class RelationshipController {

    private final RelationshipService relationshipService;

    @Autowired
    public RelationshipController(RelationshipService relationshipService) {
        this.relationshipService = relationshipService;
    }

    @RequestMapping(value = "/to/{followedId}", method = RequestMethod.POST)
    public void follow(@PathVariable("followedId") Long followedId) {
        relationshipService.follow(followedId);
    }

    @RequestMapping(value = "/to/{followedId}", method = RequestMethod.DELETE)
    public void unfollow(@PathVariable("followedId") Long followedId) throws RelationshipNotFoundException {
        relationshipService.unfollow(followedId);
    }

    @ResponseStatus(value = HttpStatus.NOT_FOUND)
    @ExceptionHandler(RelationshipNotFoundException.class)
    public void handleRelationshipNotFound() {
    }

}

File: src/main/kotlin/com/myapp/controller/RelationshipController.kt
package com.myapp.controller

import com.myapp.repository.exception.RelationshipDuplicatedException
import com.myapp.service.RelationshipService
import com.myapp.service.exception.RelationshipNotFoundException
import org.springframework.http.HttpStatus
import org.springframework.web.bind.annotation.*

@RestController
@RequestMapping("/api/relationships")
class RelationshipController(
    private val relationshipService: RelationshipService
) {

    @PostMapping(value = "/to/{followedId}")
    fun follow(@PathVariable("followedId") followedId: Long) {
        relationshipService.follow(followedId)
    }

    @DeleteMapping(value = "/to/{followedId}")
    fun unfollow(@PathVariable("followedId") followedId: Long) {
        relationshipService.unfollow(followedId)
    }

    @ResponseStatus(value = HttpStatus.BAD_REQUEST, reason = "already followed")
    @ExceptionHandler(RelationshipDuplicatedException::class)
    fun handleRelationshipDuplicated() {
    }

    @ResponseStatus(value = HttpStatus.NOT_FOUND)
    @ExceptionHandler(RelationshipNotFoundException::class)
    fun handleRelationshipNotFound() {
    }

}
